### PR TITLE
add array function auc

### PR DESCRIPTION
**Pull request by taiyang-li**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
- Add array function auc(used in machine learning). See more details: 
 https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc

### CODE REVIEW

**Comment by taiyang-li** [comment]
The Two failed integration tests seem not caused by my commit.

File: dbms/src/Functions/array/auc.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        // Calculate AUC
+        size_t curr_cnt = 0;
+        size_t curr_pos_cnt = 0;
+        size_t curr_sum = 0;
+        ResultType last_score = -1;
+        ResultType rank_sum = 0;
+        for (size_t i = 0; i < pairs.size(); ++i)
+        {
+            if (pairs[i].score == last_score)
+            {
```
Comment (sundy-li) [inline dbms/src/Functions/array/auc.cpp]: image


From the formula, we just need to calculate the sum(Rank of positives), others are constants. This code may be a little complicated.

File: dbms/src/Functions/array/registerFunctionsArray.cpp [hunk 2]
```diff
@@ -33,7 +33,7 @@ void registerFunctionArrayDistinct(FunctionFactory & factory);
 void registerFunctionArrayFlatten(FunctionFactory & factory);
 void registerFunctionArrayWithConstant(FunctionFactory & factory);
 void registerFunctionArrayZip(FunctionFactory & factory);
-
+void registerFunctionAUC(FunctionFactory &);
```
Comment (hczhcz) [inline dbms/src/Functions/array/registerFunctionsArray.cpp]: It is probably better to use an `array-` name for an array function.
Also, are there some longer names to consider?

Comment (taiyang-li) [reply dbms/src/Functions/array/registerFunctionsArray.cpp]: Already renamed "auc" to "arrayAUC".

File: dbms/src/Functions/array/array_auc.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    struct ScoreLabel
+    {
+        ResultType score;
+        UInt8 label;
+    };
+
+    static DataTypePtr getReturnType(const DataTypePtr & /* nested_type1 */, const DataTypePtr & nested_type2)
+    {
+        WhichDataType which2(nested_type2);
+        if (!which2.isUInt8())
```
Comment (qoega) [inline dbms/src/Functions/array/array_auc.cpp]: Probably isEnum8 is also a valid case.

File: dbms/src/Functions/array/array_auc.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                ++num_pos;
+            else
+                ++num_neg;
+        }
+        std::sort(pairs.begin(), pairs.end(), [](const auto & lhs, const auto & rhs) { return lhs.score < rhs.score; });
+
+        // Calculate AUC
+        size_t curr_cnt = 0;
+        size_t curr_pos_cnt = 0;
+        size_t curr_sum = 0;
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply dbms/src/Functions/array/array_auc.cpp]: I think `curr_cnt` and `curr_pos_cnt` are less than array length, for which `size_t` is enough. 
`curr_sum` may exceed `size_t`, so I replace `size_t` with `Int64` in my latest commit.

File: docs/en/query_language/functions/array_functions.md [hunk 5]
```diff
@@ -896,5 +896,11 @@ Result:
 │ [('a','d'),('b','e'),('c','f')]            │
 └────────────────────────────────────────────┘
 ```
+## arrayAUC(arr_scores, arr_labels)
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply docs/en/query_language/functions/array_functions.md]: Thanks for your link. Ialready updated.

File: docs/en/query_language/functions/array_functions.md [hunk 6]
```diff
@@ -896,5 +896,11 @@ Result:
...
+
+Returns AUC (Area Under the Curve, which is a concept in machine learning, see more details: https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc).
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply docs/en/query_language/functions/array_functions.md]: Updated in latest commit

File: dbms/tests/queries/0_stateless/01064_array_auc.sql [hunk 7]
```diff
@@ -0,0 +1 @@
+select arrayAUC([0.1, 0.4, 0.35, 0.8], [0, 0, 1, 1])
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply dbms/tests/queries/0_stateless/01064_array_auc.sql]: Already updated. 
By the way,  I use python library sklearn to check my result. For example:

``` python
import numpy as np
from sklearn import metric

def auc(scores, labels):
    y = np.array(labels)
    pred = np.array(scores)
    fpr, tpr, thresholds = metrics.roc_curve(y, pred)
    return metrics.auc(fpr, tpr) 

if __name__ == '__main__':
    scores = [0.1, 0.35, 0.4, 0.8]
    labels = [1, 0, 1, 0]
    print(auc(scores, labels))
```

File: dbms/src/Functions/array/arrayScalarProduct.h [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        // Basic type check
+        std::vector<DataTypePtr> nested_types(2, nullptr);
+        for (size_t i = 0; i < getNumberOfArguments(); ++i)
+        {
+            const DataTypeArray * array_type = checkAndGetDataType<DataTypeArray>(arguments[i].get());
+            if (!array_type)
+                throw Exception("All argument for function " + getName() + " must be an array.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply dbms/src/Functions/array/arrayScalarProduct.h]: Already updated.

File: dbms/src/Functions/array/arrayScalarProduct.h [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+            auto & nested_type = array_type->getNestedType();
+            WhichDataType which(nested_type);
+            bool is_number = which.isNativeInt() || which.isNativeUInt() || which.isFloat();
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply dbms/src/Functions/array/arrayScalarProduct.h]: Updated.

File: dbms/src/Functions/array/array_auc.cpp (hunk 3 above)
Comment (taiyang-li) [reply dbms/src/Functions/array/array_auc.cpp]: Updated

File: docs/en/query_language/functions/array_functions.md [hunk 10]
```diff
@@ -896,5 +896,11 @@ Result:
...
+
+`arr_scores` represents scores prediction model gives, while `arr_labels` represents labels of samples, usually 1 for positive sample and 0 for negtive sample.
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply docs/en/query_language/functions/array_functions.md]: As I refered to other implementations of AUC algorithm(sklearn in python,  github.com/gonum/stat in golang), labels usually only have two kinds of values, {-1, 1} and {0, 1},  and 1 as positive while -1 or 0 as negative.  

In my implementation, an exception will be throw if label values are neither from {-1, 1} nor {0, 1}.

**Comment by alexey-milovidov** [comment]
@taiyang-li I did not understand the code in the inner loop.
I tried to imagine the algorithm and have rewrite it to something more simple:
https://github.com/ClickHouse/ClickHouse/pull/9815/commits/27f0b932378c8ea2d7562a8e69404a51a1cd03d7

I'm not sure that it's 100% correct (I'm not into machine learning) but your tests have passed.

**Pull request merged**